crates/provider-http-server/src/path.rs

```diff
@@ -0,0 +1,244 @@
+//! This module contains the implementation of the `wrpc:http/incoming-handler` provider in path-based mode.
+//!
+//! In path-based mode, the HTTP server listens on a single address and routes requests to different components
+//! based on the path of the request.
+
+use core::time::Duration;
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::{bail, Context as _};
+use axum::extract::{self};
+use axum::handler::Handler;
+use axum_server::tls_rustls::RustlsConfig;
+use axum_server::Handle;
+use tokio::sync::RwLock;
+use tokio::task::JoinHandle;
+use tracing::{debug, error, info, instrument};
+use wasmcloud_provider_sdk::{HostData, LinkConfig, LinkDeleteInfo, Provider};
+
+use crate::{
+    build_request, get_cors_layer, get_tcp_listener, invoke_component, load_settings,
+    ServiceSettings,
+};
+
+/// `wrpc:http/incoming-handler` provider implementation with path-based routing
+#[derive(Clone)]
+pub struct HttpServerProvider {
+    // NOTE: When acquiring locks to both of these maps, always acquire the `paths` lock first
```

Comment:
Nit: Should we wrap these two things into a struct so we can guarantee the order of locking? At a bare minimum it can just have the two hashmaps, or you could design something different to avoid cloning the data if you felt ambitious

Reply:
Also, I am super curious why doing it in the other order causes a deadlock 😅

Reply:
>Also, I am super curious why doing it in the other order causes a deadlock 😅

It's definitely not guaranteed, but if you have two places in the code that acquire (and try to hold) the locks in the opposite order, one tokio task may acquire the first lock at the same instant that another task acquires the second lock, and then they are both holding their locks while waiting for the other one /shrug. It's not likely, I just left this message out of prudence since we've hit that before

Wrapping in a struct seems reasonable here